shared/util/ix.ts

```diff
@@ -113,3 +114,56 @@ export function asyncGeneratorFromPromise<P extends unknown[], R>(
         yield await func(...args)
     }
 }
+
+/** The maximum number of promise results to cache. */
+export const PROMISE_CACHE_CAPACITY = 5
+
+/**
+ * Memoizes a function that returns a promise. Internally, this maintains a simple
+ * bounded LRU cache.
+ *
+ * @param func The promise function.
+ */
+export function cachePromiseProvider<P extends unknown[], R>(
+    func: (...args: P) => Promise<R>,
```

Comment:
question:
can this be more typesafe? we cast `P` to `[sourcegraph.TextDocument, sourcegraph.Position]` later but my typescript fu is not good enough to know if it's easy to propogate that to this signature

Reply:
Maybe but I can't type it correctly so I'm not getting yelled at by some langserver somewhere.